Sources/MongoSwift/MongoError.swift

```diff
@@ -2,277 +2,287 @@ import CLibMongoC
 import Foundation
 
 /// An empty protocol for encapsulating all errors that this package can throw.
-public protocol MongoError: LocalizedError {}
-
-// TODO: update this link and the one below (SWIFT-319)
-/// A MongoDB server error code.
-/// - SeeAlso: https://github.com/mongodb/mongo/blob/master/src/mongo/base/error_codes.yml
-public typealias ServerErrorCode = Int
+public protocol MongoErrorProtocol: LocalizedError {}
 
 /// Protocol conformed to by errors that may contain error labels.
-public protocol LabeledError: MongoError {
+public protocol MongoLabeledError: MongoErrorProtocol {
     /// Labels that may describe the context in which this error was thrown.
     var errorLabels: [String]? { get }
 }
 
 /// Protocol conformed to by errors returned from the MongoDB deployment.
-public protocol ServerError: LabeledError {}
+public protocol MongoServerError: MongoLabeledError {}
 
-/// Thrown when commands experience errors on the server that prevent execution.
-public struct CommandError: ServerError {
-    /// A numerical code identifying the error.
-    public let code: ServerErrorCode
+/// A protocol describing errors caused by improper usage of the driver by the user.
+public protocol MongoUserError: MongoErrorProtocol {}
 
-    /// A human-readable string identifying the error code.
-    public let codeName: String
+/// The possible errors that can occur unexpectedly driver-side.
+public protocol MongoRuntimeError: MongoErrorProtocol {}
 
-    /// A message from the server describing the error.
-    public let message: String
+/// Namespace containing all the error types introduced by this driver and their dependent types.
+public enum MongoError {
+    // TODO: update this link and the one below (SWIFT-319)
+    /// A MongoDB server error code.
+    /// - SeeAlso: https://github.com/mongodb/mongo/blob/master/src/mongo/base/error_codes.yml
+    public typealias ServerErrorCode = Int
 
-    /// Labels that may describe the context in which this error was thrown.
-    public let errorLabels: [String]?
+    /// Thrown when commands experience errors on the server that prevent execution.
+    public struct CommandError: MongoServerError {
```

Comment:
We could even do sub-namespaces here for each of the error categories (e.g. `MongoError.ServerError.CommandError`), but that felt like overkill.